File: posts/transform/transform.R
# Create one file
library(arrow)
library(dplyr)

# path_to_partitioned_first_period_file <- "posts/import/data/I80/period=first/part-0.parquet"
# 
# open_dataset(path_to_partitioned_first_period_file) |> 
#   dplyr::collect() |> 
#   arrow::write_parquet(sink = "posts/transform/data/i80_period1.parquet")





# Load parquet file 
path_to_first_period_file <- "posts/transform/data/i80_period1.parquet"
data_R <- arrow::read_parquet(file = path_to_first_period_file)

data_R



# Clean dataframe names
library(janitor)
data_R <- janitor::clean_names(data_R) 




# How to create a time column?
library(lubridate)

data_R <- data_R |>
  dplyr::mutate(actual_time = lubridate::as_datetime(global_time / 1000, 
    origin = "1970-01-01",
    tz = "America/Los_Angeles"
  ))


data_R <- data_R |> 
  group_by(vehicle_id) |> 
  mutate(time = (0:(n()-1))/10) |> 
  ungroup()


# How to create variables for the preceding vehicle?

## First: Sort by Vehicle ID and Time
data_R <- data_R |> 
  arrange(vehicle_id, frame_id)

## Then create new cols
data_R <- data_R |> 
  dplyr::group_by(frame_id) |>  # grouping by frame id
  dplyr::mutate(preceding_local_y = local_y[match(preceding, vehicle_id)],
                preceding_local_x = local_x[match(preceding, vehicle_id)],
                preceding_length = v_length[match(preceding, vehicle_id)],
                preceding_width = v_width[match(preceding, vehicle_id)],
                preceding_class = v_class[match(preceding, vehicle_id)],
                preceding_vel = v_vel[match(preceding, vehicle_id)],
                preceding_acc = v_acc[match(preceding, vehicle_id)]) |> 
  dplyr::ungroup()





# How to remove undesired columns?
data_R <- data_R |>
  select(-c(total_frames, starts_with("global"), following, 
            o_zone, d_zone, int_id, section_id, direction, 
            movement), global_time) 


# How to transform multiple columns?
## metric units
data_R <- data_R |> 
  mutate(across(
    .cols = c(starts_with("local"), starts_with("v_"), space_headway, starts_with("preceding"), -preceding, -preceding_class, -v_class),
    .fns = ~ round(.x * .3048, 2)
  ))

## factor type
data_R <- data_R |> 
  mutate(across(
  .cols = c(vehicle_id, v_class, lane_id, preceding, preceding_class),
  .fns = ~ as.factor(.x)
))






# Visualization with one vehicle
data_R_veh <- data_R |> 
  dplyr::filter(vehicle_id == "2")

library(ggplot2)
ggplot(data = data_R_veh,
       mapping = aes(x = time, y = v_vel)) +
  geom_path() +
  labs(x = "Time (s)", y = "Velocity (m/s)",
       title = "Velocity of vehicle # 2000") +
  theme_minimal()



ggplot(data = data_R_veh |> filter(time == 10.0)) +
  geom_rect(aes(fill = "Preceding Veh", 
                xmin = preceding_local_y - preceding_length,
                xmax = preceding_local_y,
                ymin = preceding_local_x - (preceding_width/2),
                ymax = preceding_local_x + (preceding_width/2))) +
  geom_rect(aes(fill = "Subject Veh",
                xmin = local_y - v_length,
                xmax = local_y,
                ymin = local_x - (v_width/2),
                ymax = local_x + (v_width/2))) +
  geom_hline(yintercept = 3.6, linetype = "longdash") +
  coord_fixed(ratio=1) +
  labs(fill = NULL) +
  theme_minimal()



library(gganimate)
my.animation <- ggplot(data = data_R_veh) +
  geom_rect(aes(fill = "Preceding Veh", 
                xmin = preceding_local_y - preceding_length,
                xmax = preceding_local_y,
                ymin = preceding_local_x - (preceding_width/2),
                ymax = preceding_local_x + (preceding_width/2))) +
  geom_rect(aes(fill = "Subject Veh",
                xmin = local_y - v_length,
                xmax = local_y,
                ymin = local_x - (v_width/2),
                ymax = local_x + (v_width/2))) +
  geom_hline(yintercept = 3.6, linetype = "longdash") +
  coord_fixed(ratio=3.5) +
  labs(fill = NULL) +
  theme_minimal() + 
  theme(legend.position = "none") +
  transition_reveal(time)

animate(my.animation, height = 3,
        width = 8, units = "in", res = 300)


anim_save(filename = "posts/transform/car_following.gif")

File: posts/transform/transform.jl
# Load parquet file
# ] add Parquet
using DataFrames
using Parquet2: Dataset
path_to_first_period_file = "data/i80_period1.parquet"
data_jl = DataFrame(Dataset(path_to_first_period_file));


using TidierData
# @glimpse(data_jl)



data_jl = @clean_names(data_jl)
first(data_jl)


# https://github.com/TidierOrg/TidierData.jl/issues/1#issuecomment-1656262978
data_jl = @select(data_jl, Not(:o_zone, :d_zone, :int_id, :section_id, :direction, :movement))
count(x -> ismissing(x), data_jl.frame_id)
for col in names(data_jl)
    data_jl[!, col] .= coalesce.(data_jl[!, col], 0)
end
# @glimpse(data_jl)


using Dates

data_jl.actual_time = DateTime.(Dates.unix2datetime.(data_jl.global_time ./ 1000))
data_jl.actual_time = data_jl.actual_time .- Hour(7)


data_jl = @arrange(data_jl, vehicle_id, frame_id)
first(data_jl, 5)


data_jl = @chain data_jl begin
    @group_by(vehicle_id)
    @mutate(time = (0:(first(n()) - 1))/10)
    @ungroup()
end




data_jl = leftjoin(
           data_jl,
           data_jl[:, [:frame_id, :vehicle_id, :local_y, :v_length, :v_width, :v_class, :v_vel, :v_acc]],
           on = [:frame_id => :frame_id, :preceding => :vehicle_id],
           makeunique = true, 
           renamecols = "" => "_preceding"
       )

names(data_jl)

data_jl = @arrange(data_jl, vehicle_id, frame_id)
first(data_jl, 5)

data_jl = @chain data_jl begin
    @rename(preceding_local_y = local_y_preceding,
            preceding_length =  v_length_preceding,
            preceding_width = v_width_preceding,
            preceding_class = v_class_preceding,
            preceding_vel = v_vel_preceding,
            preceding_acc = v_acc_preceding)
end



data_jl = @select(data_jl, Not(:total_frames, starts_with("global"), :following))



# data_jl = @chain data_jl begin
#     @mutate(across((
#         starts_with("local"),
#         v_length, v_width, v_vel, v_acc,
#         space_headway,
#         preceding_local_y, preceding_length, preceding_width,
#         preceding_vel, preceding_acc
#         ), (x -> round.(x * 0.3048, digits=2))))
# end

function convert_and_round(x)
    return round(x * 0.3048, digits=2)
end

cols_to_mutate = ["local_x", "local_y", "v_length", "v_width", 
                 "v_vel", "v_acc", "space_headway",
                 "preceding_local_y", "preceding_length", "preceding_width",
                 "preceding_vel", "preceding_acc"]

for col in cols_to_mutate
    data_jl[!, col] .= convert_and_round.(data_jl[!, col])
end



data_jl[!,:vehicle_id] = string.(data_jl[!,:vehicle_id])
data_jl[!,:v_class] = string.(data_jl[!,:v_class])
data_jl[!,:lane_id] = string.(data_jl[!,:lane_id])
data_jl[!,:preceding] = string.(data_jl[!,:preceding])
data_jl[!,:preceding_class] = string.(data_jl[!,:preceding_class])



data_jl_veh = @filter(data_jl, vehicle_id == "2")



using TidierPlots
ggplot(data = data_jl_veh) +
  geom_path(aes(x = "time", y = "v_vel"), color = "blue") +
  geom_path(aes(x = "time", y = "preceding_vel"), color = "orange") +
  labs(x = "Time (s)", y = "Velocity (m/s)",
       title = "Velocity of vehicle # 2 and its preceding vehicle") +
  theme_minimal()